### ISSUE TITLE
Transaction Throughput and SNARG Efficiency

### ISSUE THREAD

**Issue by bpfarmer** [mir-core/mips#11]
The basic elements of Mir are becoming more clear: records that include predicates that must be satisfied in order for the state of a record to be updated. Predicate satisfaction is validated by checking a SNARG that demonstrates that the prover knows a witness that satisfies a record's predicate and updates the state of the record to a particular value. There are numerous advantages to this approach, some of which are: it's much less expensive than validation by reexecution, there are no gas costs: fees are based on the size of the transaction, it supports greater privacy, and it should be easier and more efficient to use SNARKs to verify entire blocks compared to using an arbitrary VM, allowing us to shrink the required state necessary for full validation.

However, this approach involves several drawbacks, including two that have kept me up at night, specifically: SNARK proving time and CRS generation. The state of SNARG/SNARK constructions is fairly immature. The tradeoffs are such that it's difficult to arrive at a proving system that provides succinct proofs, relatively fast proving times, fast verification, and post-quantum security, but doesn't require a CRS or trusted setup. SNARKs like [Groth16](https://eprint.iacr.org/2016/260.pdf) give short proof sizes and O(1) verification, but proving times quickly increase with the number of R1CS constraints, and they require a CRS.

I'd like to use this issue to start to discuss workarounds for SNARKs' long proving times, taking, as an example, a generic record maintaining the state of a large (~2^32 leaves) accumulator, specifically a Merkle prefix tree. There are a number of design concerns, including:

1. Allowing transactions to be created in the resource-constrained light client model.
2. Making updates to the record verifiable via SNARG.
3. Allowing a block producer to verifiably claim the fee associated with a user's transaction.  
4. Preserving throughput, so that multiple updates can be made to a record in a single block.

I would argue that these are all required, since supporting light clients is critical for usability, verification-via-SNARGs preserve the properties above, block producers must be incentivized to include transactions, and throughput can't limit usability or restrict parity with centralized services.

For a sketch of a possible solution, consider a light client wishing to update the state of the generic record R to include a leaf l_k. A candidate system would allow the light client to create a transaction T that specified updating R to include l_k, and included a fee designed to incentivize the inclusion of T in a block. Now, the hard part is producing a SNARG showing that a particular set of updates to R is valid and results in some new tree root. Even with a light client, producing such a SNARG would probably be possible, as long as the Merkle tree used a Pedersen or MiMC hash function instead of SHA-256, but it still requires computing a valid witness, which means that the light client needs to get the state of the record somehow.

A natural move would be to rely on block producers to produce the witness and SNARG for each update. BP's could store or query the state of R to generate a witness, and then create a SNARG. The problem then is that the number of constraints quickly becomes unmanageable within a single block, assuming that there are even 100 updates per block. If the height of the Merkle tree is ~32, and each Pedersen hash requires ~500 constraints, then each proof of inclusion for each update would require > 30000 constraints, ignoring any additional constraints placed on updates, so assuming 2^21 constraints is a natural upper bound (the current Zcash Powers of Tau limit), then the number of updates that could be batched together is fairly low. Replicating the FB user mappings is then somewhat pointless if we can only support ~10 updates per block.

Now we've arrived at the source of my worry: whether we'll inevitably run into technical limitations for SNARKs before we're actually able to provide any value. I think that we can actually get around this, at least for this case, pretty easily. Suppose that we have 1000 updates to a particular record in a block, but our circuit supports 10. The nice thing about Groth16 SNARKs is that they are **very** succinct: for the curve used in Zexe from the BLS family, the proof size is 192 bytes for 128-bit security parameters. For an obvious (in retrospect) solution, I think that we can support sequences of ordered proofs within a block, such that if we have a set of updates `u_1,...,u_n` where `n` is large, and our SNARKs allow for the batching of `k` updates at a time, we can create `k`-sized groups of updates, such that a particular group `u_{k+l},...,u_{2k+l}` takes the root of the Merkle tree for record `R` from `R_m \to R_m+1`. Then, the solution is straightforward: if the proving time to batch a group of `k` updates is `t`, then we can parallelize, so that creating proofs for all groups takes `t` wall-clock time. Then, all of these proofs (we could also do more recursion, but I don't think it's necessary), could be included in the block in sequence, such that a verifier could follow the sequence of Merkle roots from `R_1 \to R_{n/k}` and verify all 1000 updates. 

I think that in this case, depending on circuit optimization, we should still be able to beat reexecution in terms of data traveling over the network, which would be impressive, as records of this type should most favor reexecution.

Hopefully this makes sense and there are no glaring errors. The remaining issue is how to allow Block Producers to verifiably collect transaction fees. In theory, this should only be an issue for records that involve shared state and concurrent updates, as other types of records should allow users to either create their own SNARGs and submit them with a fee. Without thinking about it in detail yet, a few (admittedly vague) options are:

1. Supporting a few authenticated data types within the protocol, and verifying the associated proof-of-inclusion/exclusion within every block.
2. Giving transaction fees two components: the fee itself and then a deposit. The fee can't be retrieved by the user, but the user must assign the fee to a block producer to redeem the deposit. We rely on altruism for users to correctly assign fees to BPs.
3. Allowing a predicate to be specified within a transaction or on the record, such that the BP can create a proof satisfying that predicate (showing that an update was applied) to spend the transaction fee.

**Comment by dlubarov** [comment]
> A natural move would be to rely on block producers to produce the witness and SNARG for each update.

Just to clarify, it sounds like this idea would be used in conjunction with something resembling #3, where transactions create public update messages ("data commands" in #3) which BPs can aggregate?

> we can support sequences of ordered proofs within a block

This could definitely work, and I see the appeal of not using reexecution at all, but I feel it might be overkill since (assuming a model like #3) these updates are very uniform and inexpensive.

Judging by the data [here](https://en.bitcoin.it/wiki/Non-specialized_hardware_comparison#CPUs.2FAPUs), it looks like any "reasonable" CPU can handle at least 10^6 hashes per second. (Modern desktop or server CPUs seem to do 10^7 or more.) So if our average Merkle tree height was 32, they could handle ~30k updates per second.

30k TPS seems too high for one shard; my guess is that we'll end up with 100 TPS or less. So I don't think reexecuting Merkle operations would ever become a bottleneck; I would expect the bottleneck to be either SNARK verification (I know Groth16 verification is cheap, but probably not as cheap as ~32 hashes?), or maybe bandwidth.

**Comment by dlubarov** [comment]
Oh, did you want to avoid reexecution here not for performance reasons, but for CODA style proofs? If so that makes sense.

**Comment by bpfarmer** [comment]
So, I was working under the assumption that stateless block verification would be a desirable property. In other words, if I have block n+1, I may not know whether the initial state is valid relative to the genesis block, but I know that the state transition from n to n+1 is valid, without needing to hold onto a bunch of state. This has a lot of advantages, but also sets the foundation for actually being able to do Coda-style proofs.

> it looks like any "reasonable" CPU can handle at least 10^6 hashes per second

It's definitely the case that from a verification time perspective, reexecution is certainly the best option. However, if we want to preserve the above property, I'm more worried about bandwidth and the cost of replicating data across nodes than verification time. For updates to a binary tree of depth 32, each update proof would be 992 bytes, which is fairly large for a single leaf (it's true that we could prune some of the hashes to batch multiple updates), and then we'd possibly need to transmit additional signature or validation information for the leaf, depending on the record. The alternative with pure reexecution is that we just transmit updated values without proof, but then we break the stateless verification property

> Oh, did you want to avoid reexecution here not for performance reasons, but for CODA style proofs? If so that makes sense.

Yeah if we want to do CODA-style proofs, then we can't do pure reexecution, so we'd need to transmit proofs of inclusion in blocks, in which case we're producing the same proofs as in the proposed model. Plus, as SNARGs advance, I think that we'll see smaller proof sizes for a given number of constraints, so we'll be able to continue to beat the bandwidth requirements of transmitting proofs of inclusion.

**Comment by dlubarov** [comment]
Hm let's say we go all-in with a CODA-style protocol. I'm imagining that miners would collect a list of transactions `T`, then generate a SNARG (let's assume there's just one for now) which takes `T` as inputs, and outputs

  - Whether `T` is valid
  - The Merkle root of `T`
  - The Merkle root of the updated account set
  - The Merkle root of all updated data structures (as in #3)
  - Maybe some other Merkle roots (receipts, etc.)

And light clients would download only that SNARG. But I think other miners would need to download `T`, to update their local copies of the account set etc. (I'm assuming we use stateful miners.) Or they could download a different SNARK which outputs some kind of delta for each data structure, but I figure the deltas would still be `O(|T|)` size, so it might be best to just download `T` and reexecute any updates. I agree about not using reexecution anywhere else though.

**Comment by dlubarov** [comment]
If I understood your idea correctly, and combining it with the above model, the miner would split `T` into a series of `k`-sized "sub-blocks", compute (not in a SNARG) the state between each sub-block, and generate a SNARG in parallel for each sub-block state transition. Is that right?

I think that seems reasonable, and definitely better than one big SNARG for the whole block. I need to think more about how it compares to CODA's approach, with a binary tree of recursive SNARGs generated by different nodes.

**Comment by bpfarmer** [comment]
> But I think other miners would need to download T, to update their local copies of the account set etc. (I'm assuming we use stateful miners.) Or they could download a different SNARG which outputs some kind of delta for each data structure, but I figure the deltas would still be O(|T|) size, so it might be best to just download T and reexecute any updates. 

Yep, I think that I agree, the miners would need to download the updated state of the record key/value bindings.

> I agree about not using reexecution anywhere else though.

Cool, I think that it makes sense from a technical perspective and it makes it easier to distinguish ourselves from other distributed computation protocols that use reexecution, but obviously if there are compelling reasons to support reexecution as a mode of verification for transactions, then we should be open to those.

> If I understood your idea correctly, and combining it with the above model, the miner would split T into a series of k-sized "sub-blocks", compute (not in a SNARG) the state between each sub-block, then generate a SNARG in parallel for each sub-block state transition. Is that right?

Yep, rather than have a record's state transition be per-block, we would support sub-transitions between sub-blocks, as you say, which would allow us to preserve throughput while using SNARGs.

> I think that seems reasonable, and definitely better than one big SNARG for the whole block. I need to think more about how it compares to CODA's approach, with a binary tree of recursive SNARGs generated by different nodes.

Yeah we simply couldn't handle the proving time for constructing a single proof for even 1000 updates per block. I think that we can use the same approach as CODA if we wanted recursive proofs: we could just construct a binary tree out of all of the included proofs in a block, though maybe we'd have to group the associated sub-block proofs together. As I learn more about proofs though, I'm somewhat skeptical about the CODA approach - it's unclear how exactly they're picking their curves so that the recursive composition doesn't get crazily expensive. The Cocks-Pinch method for picking related curves only works for a single recursion, so I need to learn more about their approach.

**Comment by bpfarmer** [comment]
What are your thoughts on how deeply the storage model should be incorporated in the protocol? Should proofs of updated data structures be required or defined as features in the base protocol, or as we've discussed with sidechains, could we accomplish this with the record model? 

My thoughts on this are fairly vague and ill-defined, but I wonder if it's possible to design a scheme such that a record could require a proof from a "filecoin" record that maintained provable state replication. Basically it would just be included in the predicate that some form of proof must be provided in order to satisfy the update conditions. It's unclear whether this would be feasible to verify using a SNARG that was also verifying other conditions, but it would be nice to keep the base chain as simple as possible and not have to commit to a particular filecoin-like data-replication scheme. The downside is that it might be complex to implement and it's a bit unclear how incentives would work. Would transactions include storage fees as well as block propagation fees?

**Comment by dlubarov** [comment]
> a record could require a proof from a "filecoin" record that maintained provable state replication

Interesting, so it'd be kind of a layer 2 solution (in the sense that it wouldn't need any first-class base chain features), but using the base chain's consensus rather than a separate blockchain?

I think you're right that replication proofs could be verified by record predicates, and we could probably do regular challenges and slashing all within the Zexe model. It seems more difficult to maintain the pool of storage nodes and assign them to blobs though; we might need first-class features for that (or at least to make it easier).

**Comment by bpfarmer** [comment]
> It seems more difficult to maintain the pool of storage nodes and assign them to blobs though; we might need first-class features for that (or at least to make it easier).

I think that you're right - my intuition is that we could construct a scheme that would work, but it would be complicated.

> apparently they found a 2-cycle with all the right properties

I forget what the exact problem was but I think that a follow-up paper by the same authors presented a candidate 2-cycle of curves, but one of the curves has a low embedding degree, so I think that you'd need the order of the underlying field to be really big to provide 128-bit security, which is obviously expensive. I think that the Zexe paper goes into that as well, specifically justifying the efficiency of their approach, as they only rely on a single recursion.

> I think we could figure out a way to do it with a constant amount of state (in a record payload) though. Like the sidechain's exit receipts could be assigned sequential integers, and the base chain would only accept a receipt with the next integer.

I'll create a new issue talking about this specifically, as I think it's really interesting, but I'm fairly sure that we could do exits with a single record, which would be pretty cool.
